torch/csrc/jit/codegen/cuda/scheduler/vectorize_helper.cpp

```diff
@@ -17,171 +18,1092 @@ namespace fuser {
+        //    if (find_outer_it != ids.end() && find_inner_it != ids.end() &&
+        //      !hasPartialExtent(merge->inner())) {
+        //      Comment here.
+
+        // Remove outer
+        ids.erase(find_outer_it);
+
+        const auto inner_numerator = inner_mapping.getNumerator();
+        const auto inner_denominator = inner_mapping.getDenominator();
+
+        // Is the inner dimension is fully mapped
+        auto inner_is_fully_mapped = IrBuilder::eqExpr(
+            SimplifyingIrBuilder::divExpr(inner_numerator, inner_denominator),
+            commonOrConstExtent(ca_map_, split->inner()));
+
+        // Divisibility checks are done later, simply propagate fractional
+        // values through the graph.
+
+        // Always propagate inner dimension to in backward through split
+        auto in_mapping = inner_mapping;
+        // If inner is fully mapped propagate the outer mapping as well
+        const auto outer_mapping = getMappedExtent(split->outer());
+        in_mapping.maybeMul(inner_is_fully_mapped, outer_mapping);
+        projected_extent_[split->in()] = in_mapping;
+      } else {
+        // Only inner maps
+        projected_extent_[split->in()] = getMappedExtent(split->inner());
+        // Clear to the left of inner in since only inner maps
+        ids.erase(ids.begin(), ids.begin() + inner_pos);
+      }
+    } else if (const Merge* merge = dynamic_cast<const Merge*>(expr)) {
+      auto find_out_it = std::find(ids.begin(), ids.end(), merge->out());
+      if (find_out_it == ids.end()) {
+        continue;
+      }
+
+      auto out_pos = std::distance(ids.begin(), find_out_it);
+
+      // Map inner and outer dimensions. If outer doesn't map fully this will be
+      // found during evaluation. Don't have to worry about it here.
+      ids[out_pos] = merge->outer();
+      ids.insert(ids.begin() + out_pos + 1, merge->inner());
+
+      auto out_mapping = getMappedExtent(merge->out());
+      if (out_mapping.isZero()) {
+        ids.erase(ids.begin(), ids.begin() + out_pos + 1);
+        continue;
+      }
+      const auto out_numerator = out_mapping.getNumerator();
+      const auto out_denominator = out_mapping.getDenominator();
+
+      auto out_bigger_than_inner = IrBuilder::geExpr(
+          SimplifyingIrBuilder::divExpr(out_numerator, out_denominator),
+          commonOrConstExtent(ca_map_, merge->inner()));
+
+      // Divisibility checks are done later, simply propagate fractional
+      // values through the graph.
+
+      ProjectedExtent inner_mapping;
+      // If out is larger than the inner extent use the inner extent
+      inner_mapping.multiplyNumeratorValue(SimplifyingIrBuilder::whereExpr(
+          out_bigger_than_inner,
+          commonOrConstExtent(ca_map_, merge->inner()),
+          FusionGuard::getCurFusion()->oneVal()));
+      // otherwise use the out mapping extent
+      inner_mapping.maybeMul(
+          SimplifyingIrBuilder::notExpr(out_bigger_than_inner), out_mapping);
```

Comment:
out_mapping isn't a Val*, it's a Projected Extent. So this preserves the numerator/denominator in a way that's easier to simplify rather than having a complex Val* expression.